src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -394,43 +410,19 @@ RouterModifyTaskForShardInterval(Query *originalQuery, ShardInterval *shardInter
 	{
 		RelationRestriction *restriction = lfirst(restrictionCell);
 		List *originalBaserestrictInfo = restriction->relOptInfo->baserestrictinfo;
-		Var *relationPartitionKey = PartitionColumn(restriction->relationId,
-													restriction->index);
-		Param *uninstantiatedParameter = NULL;
+		Index rteIndex = restriction->index;
 
-		/*
-		 * We don't need to add restriction to reference tables given that they are
-		 * already single sharded and always prune to that single shard.
-		 */
-		if (PartitionMethod(restriction->relationId) == DISTRIBUTE_BY_NONE)
+		if (!allRelationsJoinedOnPartitionKey || allReferenceTables)
 		{
 			continue;
 		}
 
-		hashedOpExpressions = HashedShardIntervalOpExpressions(shardInterval);
-		Assert(list_length(hashedOpExpressions) == 2);
-
-		/*
-		 * Here we check whether the planner knows an equality between the partition column
-		 * and the uninstantiated parameter. If such an equality exists, we simply add the
-		 * shard restrictions.
-		 */
-		uninstantiatedParameter = UninstantiatedParameterForColumn(relationPartitionKey);
-		if (exprs_known_equal(restriction->plannerInfo, (Node *) relationPartitionKey,
-							  (Node *) uninstantiatedParameter))
-		{
-			RestrictInfo *geRestrictInfo = NULL;
-			RestrictInfo *leRestrictInfo = NULL;
-
-			OpExpr *hashedGEOpExpr = (OpExpr *) linitial(hashedOpExpressions);
-			OpExpr *hashedLEOpExpr = (OpExpr *) lsecond(hashedOpExpressions);
+		hashedOpExpressions = ShardIntervalOpExpressions(shardInterval, rteIndex);
 
-			geRestrictInfo = make_simple_restrictinfo((Expr *) hashedGEOpExpr);
-			originalBaserestrictInfo = lappend(originalBaserestrictInfo, geRestrictInfo);
+		hashedRestrictInfo = make_simple_restrictinfo((Expr *) hashedOpExpressions);
```

Comment:
In this part of the code (i.e., `INSERT .. SELECT `), it is still hashed restrict info.